apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/ReliableStreamReader.cs

```diff
@@ -0,0 +1,258 @@
+            return _metadata;
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public bool IsClosed { get; private set; }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public Session Session => _session;
+
+        /// <summary>
+        /// 
+        /// 
+        /// </summary>
+        public void Close()
+        {
+            lock (_sync)
+            {
+                if (IsClosed)
+                {
+                    return;
+                }
+                IsClosed = true;
+            }
+            StreamClosed?.Invoke(this, new StreamClosedEventArgs());
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public event EventHandler<StreamClosedEventArgs> StreamClosed;
+
+        void IDisposable.Dispose()
+        {
+            Close();
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public async Task<bool> HasData(CancellationToken cancellationToken)
+        {
+            return await ReliableConnect(cancellationToken);
+        }
+
+
+        private int _currentIndex;
+        private readonly IClock _clock;
+        private readonly IScheduler _scheduler;
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public async Task<Value> Next(CancellationToken cancellationToken)
+        {
+            Value result = await NextChunk(cancellationToken).ConfigureAwait(false);
+            while (result != null && _currentCall.ResponseStream.Current.ChunkedValue &&
+                   _currentIndex >= _currentCall.ResponseStream.Current.Values.Count)
+            {
+                result.ChunkedMerge(await NextChunk(cancellationToken));
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public async Task<Value> NextChunk(CancellationToken cancellationToken)
+        {
+            if (!await HasData(cancellationToken).ConfigureAwait(false))
+            {
+                return null;
+            }
+            if (_currentIndex >= _currentCall.ResponseStream.Current.Values.Count)
+            {
+                //we need to move next
+                _isReading = await ReliableMoveNext(cancellationToken);
+                _currentIndex = 0;
+                if (!_isReading)
+                {
+                    return null;
+                }
+            }
+            var result = _currentCall.ResponseStream.Current.Values[_currentIndex];
+            
+            _currentIndex++;
+            return result;
+        }
+
+        /// <inheritdoc />
```

Comment:
yes, there is a very good reason.  I had intended to log a nasty warning here because it means someone didnt dispose the object properly and it could result in a leaked session.  if done at scale, this can bring down spanner and cost a lot of money.
log added.  thanks for the reminder.